src/allmydata/test/test_iputil.py

```diff
@@ -228,3 +229,33 @@ def test_specific_port(self):
         s.close()
         port2 = iputil.listenOnUnused(tub, port)
         self.assertEqual(port, port2)
+
+
+class GcUtil(unittest.TestCase):
+    """Tests for allmydata.util.gcutil, which is used only by listenOnUnused."""
+
+    def test_gc_after_allocations(self):
+        """The resource tracker triggers allocations every 26 allocations."""
+        collections = []
+        self.patch(gc, "collect", lambda: collections.append(1))
+        for _ in range(2):
+            for _ in range(25):
+                gcutil.fileDescriptorResource.allocate()
+                self.assertEqual(len(collections), 0)
+            gcutil.fileDescriptorResource.allocate()
+            self.assertEqual(len(collections), 1)
+            del collections[:]
```

Comment:
This test failed on CI on one builder because len(collections) was 1 inside the loop where it was expected to be 0.  At first I thought maybe this was because CPython had internally decided it was time to do a full collection ... but it seems unlikely it would actually call `gc.collect` to do this?

So I wonder who else might be calling it... Since it only exists for the duration of this monkey-patch it seems weird that anyone even *could* call it.  Unless `fileDescriptorResource` is actually buggy somehow.

Reply:
It fails reliable for me with `trial -u allmydata.test.test_iputil` on the 2nd run, fwiw.  Perhaps the explanation is that `fileDescriptorResource` is shared mutable state.  So maybe that macOS CI job got to this test and there was already a file descriptor or two allocated, making it take fewer than 25 iterations to trigger.  If that's not *the* explanation it seems like at least *an* explanation.

Reply:
If you want a suggestion, I guess adding a `reset` method to `fileDescriptorResource` would let this test do something deterministic and I doubt it hurts the test suite to forget about a handful of file descriptors once or twice.

Reply:
I'll just stop using global state.